Build Action: post reminder comment to test on WordPress.com

jeherve opened the pull request:
#### Changes proposed in this Pull Request:

This introduces 2 new steps to the Build action:

- In the beginning of the build run, it checks if we need to post a test reminder comment on the PR.
    - If one is already posted, we do not need to post another.
    - If none exists, we create such a comment, with some generic instructions.
    - If one is already posted but the PR does not touch Jetpack any longer (i.e. does not touch de-Fusioned files), we delete the comment.
- Once the build is complete, the comment, if it exists, is updated with instructions on how to test.

~This posts a new comment. I think that is fine, but part of me wonders if we should instead edit the big comment that gets posted by the Repo gardening action.~ This was discussed below: https://github.com/Automattic/jetpack/pull/25310#pullrequestreview-1054416035

> **Note**
> Following the discussion at p9dueE-5yg-p2, I've introduced a new file, `.github/actions/build/de-fusioned-files.js`, where we'll keep a list of all de-fusioned files as we go through the project. We'll use that list to figure out if a test reminder comment needs to be posted on PRs. When Fusion is no more, we can retire that list, and instead rely on the output of the `changed` step just like we did early on this PR.

> **Note**
> You'll note some debug statements in the code. Those will not be displayed in the action output by default, but can be displayed if one enables debugging mode for GitHub actions as explained [here](https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging). Here is how the debug statements can look like:

<img width="962" alt="image" src="https://user-images.githubusercontent.com/426388/183912252-ee68c01d-bfe5-4d37-980e-c67b6a81096c.png">

**To-do**

- [x] If we go with this approach, we would probably want to limit the comment to PRs that touch the Jetpack plugin. I think we may be able to achieve that by moving the step that generates the first comment a bit later in the Build process, after the "Detect changed projects" step, so it can get the projects that are touched, and only comment if Jetpack is concerned.
- [x] We need to ensure that the comment only gets posted once, even if the build workflow is triggered multiple times (like when someone pushes to the existing branch). A quick and dirty solution would be to check for the `opened` event type in addition to the `pull_request` event name before we post the comment, but this would not support folks that open a PR with no changes to the Jetpack plugin, and then later on update their PR with new changes that touch the plugin. 

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

Internal reference: p1658931404733059-slack-C03QNBQKG73

#### Does this pull request change what data or activity we track or use?

* No

#### Testing instructions:

* Here is how the comment below looked like immediately after opening the PR:

<img width="967" alt="image" src="https://user-images.githubusercontent.com/426388/181575588-db016129-3124-431e-9df2-f991d1921414.png">

* Here is how it looks once the build has completed:

<img width="988" alt="image" src="https://user-images.githubusercontent.com/426388/181579042-21f7b937-c452-4fbf-8122-0937ff58f8e7.png">

You can see the tests here:
- A comment is posted when a de-fusioned file (`projects/plugins/jetpack/_inc/genericons.php`) is edited: #25313
- No comment is posted when the PR touches a Fusioned file: #25314

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - 0-as-1202637881054088

jeherve committed:
Build Action: post reminder comment to test on WordPress.com

This introduces 2 new steps to the Build action:

- In the beginning of the build run, it posts a comment on the PR telling you that your PR will need to be tested on WordPress.com.
- Once the build is complete, the comment is updated with instructions on how to test.

Internal reference: p1658931404733059-slack-C03QNBQKG73

jeherve committed:
Try to only trigger comment when the Jetpack plugin is concerned

anomiex commented on .github/workflows/build.yml:
```diff
@@ -96,6 +111,23 @@ jobs:
+        with:
+          script: |
+            const { BRANCH_NAME } = process.env;
+            const commentBody = `You can now test your Pull Request on WordPress.com. On your sandbox, run \`bin/jetpack-downloader test jetpack ${ BRANCH_NAME }\` to get started.`;
```
Should we adjust the wording to not assume the reader is an Automattician? Maybe just "If you are an Automattician, you can now test ..."

sdixon194 replied:
Agreed

jeherve replied:
Good call. Done in bf586a55f799967187718129d1ab9e112dc65370

anomiex commented on .github/workflows/build.yml:
```diff
@@ -96,6 +111,23 @@ jobs:
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: commentBody,
+              comment_id: ${{steps.set-reminder-comment.outputs.result}},
```
This should probably be passed through via an environment variable, like you're doing with `BRANCH_NAME`. Even though it should always be an integer, it's probably better to follow good patterns.

jeherve replied:
Done in bf586a55f799967187718129d1ab9e112dc65370

anomiex reviewed (commented):
> This posts a new comment. I think that is fine, but part of me wonders if we should instead edit the big comment that gets posted by the Repo gardening action.

I think it's fine too. Editing the comment would require both actions take care not to step on each other, and we'd have to worry about races.

> We need to ensure that the comment only gets posted once, even if the build workflow is triggered multiple times (like when someone pushes to the existing branch). A quick and dirty solution would be to check for the opened event type in addition to the pull_request event name before we post the comment, but this would not support folks that open a PR with no changes to the Jetpack plugin, and then later on update their PR with new changes that touch the plugin.

Fetch the comments and edit it if one is found, like we do in the repo gardening action?

You could make the searching for the matching comment easy by including a comment in the markdown (`<!-- like this -->`) instead of having to match the visible text.

sdixon194 reviewed (commented):
This is perfect, I think this is exactly what I was looking for! My only lingering questions if this will still get posted on PRs that need to be fusioned, since we'd only want one or the other depending on if things have been de-fusioned or not.

anomiex commented:
> My only lingering questions if this will still get posted on PRs that need to be fusioned, since we'd only want one or the other depending on if things have been de-fusioned or not.

But what if the PR touches multiple things, some that are de-Fusioned and some that aren't yet?

sdixon194 commented:
> But what if the PR touches multiple things, some that are de-Fusioned and some that aren't yet?

Well in that case we'd want both :D 

EDIT: thinking more about how this would work, I'd guess you'd have to run the script, then cherry-pick the fusion diff?

jeherve committed:
Add new step to check for an existing comment before to post one.

jeherve commented:
> Fetch the comments and edit it if one is found, like we do in the repo gardening action?

Yes, this is what I had in mind as well. I was wondering if there was a shorter way to do it, but that seems like the cleanest way. I've just committed bf586a55f799967187718129d1ab9e112dc65370 to do that.

Since I've also changed the comment body, we still have a new comment posted on this PR. I've rebased my 2 test PRs and added additional commits to better test the new step there.

anomiex commented on .github/workflows/build.yml:
```diff
@@ -48,6 +48,54 @@ jobs:
+              number: context.issue.number,
+            };
+            const comments = await github.graphql(query, variables);
+            let hasComment = false;
+            comments.repository.pullRequest.comments.nodes.map( comment => {
+              if ( comment.body.includes( '<!-- wpcom-reminder-comment -->' ) ) {
+                hasComment = true;
+              }
+            } );
+
+            return hasComment;
```
Could simplify this part, I think
```suggestion
            return comments.repository.pullRequest.comments.nodes.some( comment => comment.body.includes( '<!-- wpcom-reminder-comment -->' ) );
```

jeherve replied:
👍 done in b2e789005a734e8ee57de9d5dfc954c97b89f689

anomiex commented on .github/workflows/build.yml:
```diff
@@ -48,6 +48,54 @@ jobs:
           CHANGED="$(EXTRA=build .github/files/list-changed-projects.sh)"
           echo "::set-output name=projects::${CHANGED}"
 
+      - name: Check if a WordPress.com test reminder comment exists
```
Any particular reason for this to be a separate step instead of being merged with the next one?

jeherve replied:
No particular reason. I just thought it was more readable that way.

anomiex reviewed (commented):
Seems sensible. A few minor suggestions inline.

I wonder if we should go one step further with it and delete an existing comment if it no longer builds Jetpack.

To make that happen, you'd probably want to keep the "find existing comment" code more like you have it, but also fetch the `databaseId` field and record it.

anomiex committed:
Simplify loop through comments

See https://github.com/Automattic/jetpack/pull/25310#discussion_r933351230

jeherve committed:
Update logic to delete reminder comments when a PR is updated

See https://github.com/Automattic/jetpack/pull/25310#pullrequestreview-1055682291

This updates the logic:

- When checking for an existing comment, we'll now store the IDs of all comments that may be reminder comments (it will store multiple comments if for some reason there is more than one).
- It adds a new step that will delete any stored comments if the build action detects that the PR is no longer touching the Jetpack project.

jeherve commented:
> I wonder if we should go one step further with it and delete an existing comment if it no longer builds Jetpack.
> 
> To make that happen, you'd probably want to keep the "find existing comment" code more like you have it, but also fetch the databaseId field and record it.

I took a stab at this in 88f109ac7e502f7a7cac3ea3bffd386ed676e1ed. Let me know if that's what you had in mind.

We'll also see what happens with the last comment on this PR (the first one didn't have the markup we're looking for yet), that should be a good test.
**Update**: looks like that didn't work:
https://github.com/Automattic/jetpack/runs/7582840924?check_suite_focus=true

I'm about done for today, but I'll take a closer look next week.

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/build-reminder-comment

jeherve committed:
Remove extra space

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/build-reminder-comment

jeherve committed:
Add new function to look for all test reminder comments

We move this to a separate file so it's easier to read and maintain later on.

jeherve committed:
Add new file for our list of De-Fusioned files so far

We'll keep that list up to date as we move forward with the project.

jeherve committed:
Add main function used to check if a reminder comment is needed

This will leverage:

- the de-fusioned list
- the function used to check for existing reminder comments.

We'll call this function from the Build workflow directly, so we can check, delete, and post a test reminder comment (that will later be updated with more detailed testing instructions) on the PR, if needed (i.e. if the PR touches de-fusioned files).

jeherve committed:
Update build workflow to point to new external function

jeherve committed:
Add the files that are already de-fusioned

sdixon194 reviewed (dismissed):
This is awesome - did some tests on your PRs and it even posts both messages if files that are both fusioned and defusioned are touched. I think this is pretty robust. I think we should give it a go, but if someone more versed in github actions wants to review it as well that'd be cool! 👍

jeherve committed:
Add more de-fusioned files

jeherve committed:
More de-fusioned files

kraftbj reviewed (approved):
@jeherve land whenever you're ready!
